### ISSUE TITLE
[0.4.5] array-callback-return reports exhaustive switch callbacks whose branches all return or throw

### ISSUE THREAD

**Issue by fireairforce** [utooland/utoo-lint#1275]
## Summary

An array callback with a switch that includes default is incorrectly reported even when every branch returns a value or throws. An equivalent if/else and a switch followed by an unconditional return are accepted. A switch without default is correctly reported.

Verified with @utoo/lint 0.4.5, ESLint 8.57.1, @typescript-eslint/parser and eslint-plugin 6.21.0, eslint-plugin-react 7.37.5, Node.js 22.23.1 on macOS arm64. Each example enables only the indicated rule, with no parsing failures or suppression comments.

## Setup

```sh
npm init -y
npm install --save-dev @utoo/lint@0.4.5 eslint@8.57.1 @typescript-eslint/parser@6.21.0 @typescript-eslint/eslint-plugin@6.21.0 eslint-plugin-react@7.37.5 typescript@5.3.3
```

`tsconfig.json`:

```json
{"compilerOptions":{"strict":true,"target":"ES2022","jsx":"react"},"include":["example.tsx"]}
```

Run each case separately, replacing example.tsx and both config files with the corresponding contents below:

```sh
./node_modules/.bin/eslint --no-eslintrc --config .eslintrc.cjs --format=json example.tsx
./node_modules/.bin/utlint --config utlint.config.json --format=json example.tsx
```

## switch-all-return

`example.tsx`:

```tsx
export const result=[1].map(x=>{switch(x){case 1:return x;default:return 0;}});
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [],
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 1 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
Expected to return a value in array callback.
```

## switch-throw-default

`example.tsx`:

```tsx
export const result=[1].map(x=>{switch(x){case 1:return x;default:throw new Error();}});
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [],
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 1 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
Expected to return a value in array callback.
```

## switch-missing-default

`example.tsx`:

```tsx
export const result=[1].map(x=>{switch(x){case 1:return x;}});
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [],
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 1 diagnostic(s).

eslint messages:

```text
Array.prototype.map() expects a value to be returned at the end of arrow function.
```

utlint messages:

```text
Expected to return a value in array callback.
```

## switch-trailing-return

`example.tsx`:

```tsx
export const result=[1].map(x=>{switch(x){case 1:return x;}return 0;});
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [],
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```

## if-all-return

`example.tsx`:

```tsx
export const result=[1].map(x=>{if(x)return x;else return 0;});
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "plugins": [],
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "array-callback-return": [
      "error",
      {
        "allowImplicit": true,
        "checkForEach": false,
        "allowVoid": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```
